tmt/result.py

```diff
@@ -187,7 +187,24 @@ class CheckResult(BaseResult):
     event: CheckEvent = field(
         default=CheckEvent.BEFORE_TEST,
         serialize=lambda event: event.value,
-        unserialize=CheckEvent.from_spec)
+        unserialize=CheckEvent.from_spec
+        )
+
+    def interpret_check_result(self, interpret: CheckResultInterpret) -> 'CheckResult':
+        """Interpret check result according to the check_result interpretation."""
+
+        if interpret == CheckResultInterpret.RESPECT:
+            return self
+        if interpret == CheckResultInterpret.INFO:
+            self.result = ResultOutcome.INFO
+
+        elif interpret == CheckResultInterpret.XFAIL and self.event != CheckEvent.BEFORE_TEST:
```

Comment:
I was thinking about adding a logic that would check for both before and after and make the xfail 'pass' only if after or both are failed, but wasn't sure if it's needed. ?

Reply:
On second thought, this does need to be improved.  
So, what are the expectations?
before fails > pass
before pass > pass
after fails > pass
after fails > fail
?

Reply:
No idea, TBH. Should there be a difference between before/after treatment? Current checks do some preparation in their `before` execution, it would be hard for them to fail in `before` but pass in `after`. It seems to me the `after` should have more effect on the result outcome, it's usually where the test happens, although I wouldn't let `fail` in `before` slip as `pass`. Maybe `warning`?

@psss ^^

Reply:
I've tried to find a way to aggregate the before and after check results, but it feels a bit ugly and wrong:
```
diff --git a/tmt/result.py b/tmt/result.py
index 94c43443..19d67c2d 100644
--- a/tmt/result.py
+++ b/tmt/result.py
@@ -197,8 +197,8 @@ class CheckResult(BaseResult):
             return self
         if interpret == CheckResultInterpret.INFO:
             self.result = ResultOutcome.INFO
-
-        elif interpret == CheckResultInterpret.XFAIL and self.event != CheckEvent.BEFORE_TEST:
+        # Note that xfail applies to both BEFORE_TEST and AFTER_TEST
+        elif interpret == CheckResultInterpret.XFAIL:
             self.result = {
                 ResultOutcome.FAIL: ResultOutcome.PASS,
                 ResultOutcome.PASS: ResultOutcome.FAIL
@@ -367,10 +367,38 @@ class Result(BaseResult):
         original_result = self.result
 
         # Interpret check results
-        self.check = [
-            check_result.interpret_check_result(CheckResultInterpret(interpret_checks[check_result.name]))
-            for check_result in self.check
-            ]
+        interpreted_results: list[CheckResult] = []
+        xfail_failures = {
+            CheckEvent.BEFORE_TEST: False,
+            CheckEvent.AFTER_TEST: False
+        }
+
+        # First pass: interpret results and track XFAIL failures
+        for check_result in self.check:
+            interpretation_type = interpret_checks[check_result.name]
+            interpreter = CheckResultInterpret(interpretation_type)
+            interpreted_result = check_result.interpret_check_result(interpreter)
+
+            # Track if BEFORE/AFTER test XFAILs have failed
+            if (interpreter == CheckResultInterpret.XFAIL and
+                interpreted_result.result == ResultOutcome.FAIL and
+                check_result.event in xfail_failures):
+                xfail_failures[check_result.event] = True
+
+            interpreted_results.append(interpreted_result)
+
+        # Second pass: We want to make both xfail before and after check results pass
+        # if either or both failed
+        if any(xfail_failures.values()):
+            for _, (check_result, interpreted_result) in enumerate(
+                zip(self.check, interpreted_results)
+            ):
+                if (interpret_checks[check_result.name] == CheckResultInterpret.XFAIL and
+                    check_result.event in xfail_failures and
+                    interpreted_result.result == ResultOutcome.FAIL):
+                    interpreted_result.result = ResultOutcome.PASS
+
+        self.check = interpreted_results
 
         if interpret not in (ResultInterpret.RESPECT, ResultInterpret.XFAIL):
             self.result = ResultOutcome(interpret.value)
```

Would it make sense to only allow "xfail" for after test check? i.e. what's in my previous comment.
@happz @psss

Reply:
* `interpretation_type` should already be `CheckResultInterpret`, it should not be necessary to perform the `CheckResultInterpret(interpretation_type)` call
* I don't understand `enumerate()` when the index is thrown away via `_`

The rest, well, it's not pretty, but I'm not sure if it even *can be* pretty. It's complicated stuff, needs to trickle results through several layers. Maybe we could simplify with some functional magic, `reduce`, `map`, `any` & `all`, but I would prefer being more verbose here and now, so we all understand what's happening. We can always simplify things later if needed. LGTM as it is.

Reply:
> yeah, and I wouldn't ignore the before event, i.e. if it fails, the check fails in general.  

It wouldn't be ignored. Only when interpret is xfail and original result of before_test check is not "FAIL", we could convert it to warn, or info.

Reply:
@happz I've really tried to make the logic work, but in the end, I'd very much prefer to have any potential smarter way to do this once we have support for multiple checks per test, as it eventually breaks down having only `check.how` to go with.

Reply:
Agreed.